src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -332,12 +332,36 @@ private int commonPrefixLength(List<String> classNames) {
     for (int i = 0; i < size; i++) {
       String a = classNames.get(i);
       String b = typeSpecStack.get(i).name;
-      if (!a.equals(b)) return i;
+      if (!a.equals(b)) {
+        for (; i > 0; i--) {
+          // Does the current type have a parent type with a member type with the same name
+          if (!parentTypeHasMemberType(classNames.get(i), i)) {
```

Comment:
I don't think this is wrong, I'm trying to find the least-qualified non-conflicting name here.

I changed the comment and commit description to make that more clear.